core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java

```diff
@@ -0,0 +1,104 @@
+package io.parsingdata.metal;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import static io.parsingdata.metal.Shorthand.con;
+import static io.parsingdata.metal.Shorthand.def;
+import static io.parsingdata.metal.Shorthand.last;
+import static io.parsingdata.metal.Shorthand.opt;
+import static io.parsingdata.metal.Shorthand.ref;
+import static io.parsingdata.metal.Shorthand.repn;
+import static io.parsingdata.metal.Shorthand.seq;
+import static io.parsingdata.metal.Shorthand.tie;
+import static io.parsingdata.metal.Shorthand.token;
+import static io.parsingdata.metal.data.ParseState.createFromByteStream;
+import static io.parsingdata.metal.util.EnvironmentFactory.env;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.Test;
+
+import io.parsingdata.metal.data.ImmutableList;
+import io.parsingdata.metal.data.ParseState;
+import io.parsingdata.metal.data.ParseValue;
+import io.parsingdata.metal.data.Selection;
+import io.parsingdata.metal.expression.value.Value;
+import io.parsingdata.metal.token.Token;
+import io.parsingdata.metal.util.InMemoryByteStream;
+
+class ImmutableObjectTest {
+
+    @Test
+    void checkHashCode() {
+        final ImmutableObject immutableObject = new ImmutableObject() {
+            @Override
+            public String toString() {return "test";}
+            @Override
+            public boolean equals(Object obj) {return false;}
+            @Override
+            public int immutableHashCode() {return 25;}
+        };
+        assertEquals(25, immutableObject.hashCode());
+    }
+
+    @Test
+    void calculateHashOnlyOnce() {
+        final AtomicInteger counter = new AtomicInteger(0);
+        final ImmutableObject immutableObject = new ImmutableObject() {
+            @Override
+            public String toString() {return "";}
+            @Override
+            public boolean equals(Object obj) {return false;}
+            @Override
+            public int immutableHashCode() {
+                counter.incrementAndGet();
+                return 15;
+            }
+        };
+        assertEquals(15, immutableObject.hashCode());
+        assertEquals(15, immutableObject.hashCode());
+        assertEquals(15, immutableObject.hashCode());
+        assertEquals(1, counter.get());
+    }
+
+    @Test
+    void performanceTest() {
+        // This test would take way too much time without hash caching.
```

Comment:
Ah, `@Timeout`: https://www.baeldung.com/java-junit-5-timeout-annotation

Maybe add it, to make sure we never have a regression here?

Reply:
I did add that, but I have removed it again for two reasons:
1. The timing specified is an absolute timing and this test is dependent on the agent on GitHub, which could fluctuate. So that number would be arbitrary.
2. The TimeOut annotation does not stop the test. It simply checks if the time it took exceeds the timeout value. That means that the run would still run a very long time in case the hash cache is turned off and that would most likely trigger a GitHub error sooner than it would trigger the timeout error.

Reply:
Discussed it again with @jvdb. I'm going to tune the numbers in this test, such that without hash cashing it would take e..g 30 seconds - 1 minute to run, while with hash cashing in a few milliseconds. I could then add a TimeOut annotation with a limit of e.g. 10 seconds. Then we would not have an error build when there is a GitHub glitch or something, and the test would also not run for a very long time in case hash cashing is (accidentally) turned off.